### ISSUE TITLE
Instantiation error on android test but native ios passes

### ISSUE THREAD

**Issue by plusmobileapps** [Kodein-Framework/MocKMP#4]
I was trying to write a test using MocKmp and [Decompose](https://github.com/arkivanov/Decompose) and I can't figure out why the Android test is failing, but the iOS one is passing. Below is the interface being mocked and a simple test. 

```kotlin
internal interface RootBlocChildBlocProvider {

    fun createBottomNavigationBloc(
        context: ComponentContext,
        output: (BottomNavigationBloc.Output) -> Unit
    ): BottomNavigationBloc

}

internal class RootBlocTest : TestsWithMocks() {

    override fun setUpMocks() = injectMocks(mocker)

    @Mock lateinit var rootResultUseCase: RootResultUseCase
    @Mock lateinit var blocProvider: RootBlocChildBlocProvider
    @Mock lateinit var bottomNavigationBloc: BottomNavigationBloc
    @Mock lateinit var activityLoggerBloc: ActivityLoggerBloc
    
    private val lifecycle = LifecycleRegistry().apply { resume() }

    val root: RootBloc by withMocks {
        RootBlocImpl(
            componentContext = DefaultComponentContext(lifecycle),
            rootResultUseCase = rootResultUseCase,
            blocProvider = blocProvider,
        )
    }

    @Test
    fun WHEN_created_THEN_BottomNavigationBloc_displayed() {
       //TODO invoke outputs for other tests
        val bottomNavOutputs = mutableListOf<(BottomNavigationBloc.Output) -> Unit>()

       // must be mocked before creation since done in the init
      // error is happening at this line when mocking
        every { blocProvider.createBottomNavigationBloc(isAny(), isAny(capture = bottomNavOutputs)) } returns bottomNavigationBloc

        val root = root

        assertTrue(root.activeChild is Child.BottomNavigation)
    }
}
```

From looking at the crash log, it seems to be caused by the `ComponentContext` although that was just an interface so was slightly confused. Not sure if this is a bug, misconfigured test, or maybe the usage of [DefaultComponentContext](https://github.com/arkivanov/Decompose/blob/master/decompose/src/commonMain/kotlin/com/arkivanov/decompose/DefaultComponentContext.kt)? Any help would be appreciated.

```
com.arkivanov.decompose.ComponentContext
java.lang.InstantiationError: com.arkivanov.decompose.ComponentContext
	at jdk.internal.reflect.GeneratedSerializationConstructorAccessor11.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:48)
	at org.objenesis.ObjenesisBase.newInstance(ObjenesisBase.java:73)
	at org.kodein.mock.PlatformJvmKt.unsafeValue(platformJvm.kt:46)
	at org.kodein.mock.ArgConstraintsBuilder.toReturn(ArgConstraintsBuilder.kt:37)
	at com.plusmobileapps.wolfpack.RootBlocTest$WHEN_created_THEN_BottomNavigationBloc_displayed$1.invoke(RootBlocTest.kt:113)
	at com.plusmobileapps.wolfpack.RootBlocTest$WHEN_created_THEN_BottomNavigationBloc_displayed$1.invoke(RootBlocTest.kt:53)
	at org.kodein.mock.Mocker.every(Mocker.kt:169)
	at org.kodein.mock.tests.TestsWithMocks.every(TestsWithMocks.kt:54)
	at com.plusmobileapps.wolfpack.RootBlocTest.WHEN_created_THEN_BottomNavigationBloc_displayed(RootBlocTest.kt:53)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:121)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:832)


```

**Comment by glureau** [comment]
Looks like com.arkivanov.decompose.ComponentContext is an interface, so it may be related to https://github.com/Kodein-Framework/MocKMP/issues/7

As a workaround, you could try to use `@UsesMock(..., ComponentContext)` so that it generates a mock for this interface, and then you can replace `isAny()` by `isAny<MockComponentContext>()`.

**Comment by plusmobileapps** [comment]
That issue does sound related here, although `DefaultComponentContext` is needed for this integration test as I don't really want to use a mock of it since it instantiates routers for the class under test. After looking at your suggestion, I did manage to get it to work by actually replacing it with `isAny<DefaultComponentContext>()` so appreciate the guidance. 🚀 

---

Since #7 suggests there is a limitation to interfaces with objenesis, I am guessing it's not possible for MocKmp to ever detect the argument is a concrete class so `isAny()` could be used instead of declaring the type explicitly?

**Comment by glureau** [comment]
I think isAny, isNull and isNotNull don't care at all about the generated value, so generating a class (from mocks or somewhere else) should be ok. As I suppose it's only impacting JVM, we could rely on kotlin-reflect or java.lang.reflect Proxy ([for example](https://stackoverflow.com/a/1082869/1898596)) to generate an instance on the fly.

The 4 other verification methods requires an instance if I'm right, so I suppose we don't need to generate the class.

**Comment by SalomonBrys** [comment]
Fixed in 1.3.0.

https://github.com/Kodein-Framework/MocKMP/releases/tag/v1.3.0
